[WIP] Rewrite cam.js in typescript in separate branch

agsh opened the pull request:
I just wanted to rewrite it so that everyone would be comfortable

agsh committed:
chore: onvif type improvements

agsh committed:
chore: generated documentation

agsh committed:
chore: improve types

agsh committed:
feat: getServiceCapabilities

agsh committed:
feat: getVideoSources

agsh committed:
chore: specs update

agsh committed:
feat: getVideoSourceConfigurations

agsh committed:
feat: getVideoSourceConfigurationOptions

agsh committed:
feat: getVideoEncoderConfigurations

agsh committed:
chorus: update checklist and docs

agsh committed:
chorus: update checklist

momoAmch committed:
SetNTP is not working with DNS due to a missing schema version URL

SetNTP is not working with DNS due to a missing schema version URL

RogerHardiman committed:
Merge pull request #333 from momoAmch/patch-2

SetNTP is not working with DNS due to a missing schema version URL

mrhyde commented:
Hey! Just wanted to say thanks for all the amazing work you've put into rewriting the ONVIF library in TypeScript. It's been super helpful as a starting point.

Since this PR has been hanging around for a while and folks are really wanting a TypeScript version, I went ahead and made a stable fork based on your work. You can find it as `@2bad/onvif` on npm. I kept all the original code from here and added some stability fixes and tests while providing a stable build.

If anyone needs a working TypeScript ONVIF package right now, just run:
```sh
npm install @2bad/onvif
```

Hope this helps out the community! And hey, I'm totally down to work together on bringing these efforts together in the future.

agsh committed:
fix: add namespaces to `setNTP`

agsh commented:
@mrhyde Hello! Great work! Glad that someone combine all the things that I wrote to a working version!

The main goal of the `v1.x` is to keep the best of `0.x`, add types (I generate it from the ONVIF WSDL schemas) and make it a bit more modular.

We've got a lot of patches in `0.x` to support as much cameras as we can. Not all of them, because the problem is in manufacturers, not specification. So I need to gracefully move all the methods. 

BTW, here is my roadmap: https://github.com/agsh/onvif/blob/v1/CHECKED.md

Have you added any new onvif methods to the code?
And your PRs are welcome any time! :+1:

agsh committed:
feat: setVideoSourceConfiguration

agsh committed:
chore: npm audit fix

agsh committed:
feat: Digest Auth

rodrigograca31 commented:
here just to say that Im looking forward to this!

agsh committed:
feat: first tests in jest

agsh committed:
feat: parseSOAP test

agsh committed:
chore: update interfaces with new specs

agsh committed:
fix: some type fixes

agsh committed:
feat: real tests with happytime-onvif-server

agsh committed:
chore: github actions

agsh committed:
chore: github action to test on v1

agsh committed:
fix: fix peer dependencies and package-lock.json

agsh committed:
fix: test ip address fix

agsh committed:
fix: test add timeout to start server

agsh committed:
chore: change test server port to 8000

agsh committed:
feat: ContinuousMove and Duration type improvements

agsh committed:
feat: cleanup unnecessary interfaces in PTZ an use the same from the specs

agsh committed:
fix: remove Capabilities interface from devicemgmt

mrhyde commented:
@agsh I'd like to point a critical issue that prevents authentication with cameras

When authenticating, the current implementation checks for the presence of date/time components but doesn't properly handle zero values. If hour, minute, or second is 0, the code treats these as falsy values and throws an "Invalid date or time structure" error, preventing successful authentication.

I've implemented a fix in my branch and have tested it extensively in a production environment for the past couple of months. Due to the significant number of changes in my fork, I can't simply submit this as a standard PR. You can check out the specific commit that addresses this issue here: https://github.com/2BAD/onvif/commit/c1b5c2ba8719d5125c3ae1d2a1ae2e220b48944d

agsh commented:
> @agsh I'd like to point a critical issue that prevents authentication with cameras
> 
> When authenticating, the current implementation checks for the presence of date/time components but doesn't properly handle zero values. If hour, minute, or second is 0, the code treats these as falsy values and throws an "Invalid date or time structure" error, preventing successful authentication.
> 
> I've implemented a fix in my branch and have tested it extensively in a production environment for the past couple of months. Due to the significant number of changes in my fork, I can't simply submit this as a standard PR. You can check out the specific commit that addresses this issue here: [2BAD/onvif@c1b5c2b](https://github.com/2BAD/onvif/commit/c1b5c2ba8719d5125c3ae1d2a1ae2e220b48944d)

@mrhyde Hi! Thanks for your interest in this project!
Yes, you are right, the code you point to is buggy, but there is one problem... It is not my code. It is not in this library. It is your changes... :laughing:

https://github.com/2BAD/onvif/commit/c35c03214a4fded235f280672f1dd64e8bfd6721#diff-be0ad991d767e285d4a7362e8c021924de8bc914d42742fec462a4fa0a7769e6L279-R318

agsh committed:
feat: return whole SystemDateTime object in `getSystemDateAndTime` and `setSystemDateAndTime`, `getOnlySystemDateAndTime` method

agsh committed:
chore: compatibility layer development frozen

agsh committed:
feat: common and connection tests

agsh committed:
fix: renamed test fix

agsh committed:
feat: common tests

mrhyde commented:
> @mrhyde Hi! Thanks for your interest in this project! Yes, you are right, the code you point to is buggy, but there is one problem... It is not my code. It is not in this library. It is your changes... 😆
Oh man this is awkward 😆
I was pretty sure that code was coming from the upstream but apparently it was my oversight during type fixes. Sorry for wasting your time on this one!

agsh committed:
feat: `getCapabilities` change returned type, README update

agsh committed:
fix: compatibility fix

agsh committed:
feat: `getDeviceInformation` and `getHostname` tests

agsh committed:
chore: thanks to HappyTimeSoft for permission to use their onvif server for testing

agsh committed:
feat: `getScopes` tests, typo fix

agsh committed:
chore: cov badge

agsh committed:
chore: cov badge

agsh committed:
fix: change attributes to camelCase, not only tags

agsh committed:
feat: `setDNS` method, bugfixes, typefixes

agsh committed:
feat: device.ts tests

agsh committed:
fix: compatibility type fix

agsh committed:
chore: fresh interfaces

agsh committed:
feat: `setNetworkInterfaces` method, tests

agsh committed:
feat: slice of the work (new interfaces, tests, reworked ptz, device, onvif modules)

agsh committed:
chore: rename compatibility class (suspended work for the first version) to prevent tests failing

agsh committed:
feat: ptz new methods, types an tests

agsh committed:
feat: ptz tests, update optional fields in interfaces

agsh committed:
feat: more ptz tests

agsh committed:
feat: media methods

agsh committed:
chore: remove mocha and nyc

agsh committed:
feat: split media to different profile classes

agsh committed:
feat: small util function struct

agsh committed:
feat: media1 profile methods

agsh committed:
chore: eslint prefer-template

agsh committed:
feat: media2 methods (createProfile, addConfiguration), tests

agsh committed:
fix: empty nodes with line-break compatibility

agsh committed:
feat: media2.removeConfiguration method

agsh committed:
feat: media2.removeConfiguration method

agsh committed:
Merge remote-tracking branch 'origin/v1' into v1

agsh committed:
feat: media2.deleteProfile method

agsh committed:
chore: tests in sequence (media profiles sometimes conflicting)

agsh committed:
chore: cleanup

agsh committed:
feat: media1 add<Name>Configuration methods

agsh committed:
chore: generate cov-badge only once in the matrix

agsh committed:
feat: media1 remove<Name>Configuration methods

agsh committed:
feat: media1 getProfile uses media2 getProfiles when possible

agsh committed:
chore: documentation, dependencies

agsh committed:
chore: interfaces update

agsh committed:
feat: ptz `getConfiguration method`

agsh committed:
feat: media2 `get<Entity>Configurations` methods

agsh committed:
feat: media getAudioOutputs and `get<Entity>Configurations` methods

agsh committed:
chore: classes docs

agsh committed:
feat: media `getCompatible<Entity>Configurations` methods

agsh committed:
feat: media `getCompatible<Entity>Configuration` methods

agsh committed:
feat: media `getCompatible<Entity>ConfigurationOptions` methods

agsh committed:
chore: half way there 🚀

agsh committed:
feat: media1 `setVideoSourceConfiguration` method

agsh committed:
feat: media1 `setVideoEncoderConfiguration` method, remove media2 methods from media1 library

agsh committed:
chore: move xml object generation under namespace

agsh committed:
feat: `setVideoEncoderConfiguration` tests

agsh committed:
feat: media `setAudioSourceConfiguration`, `setAudioEncoderConfiguration` methods

agsh committed:
chore: update interfaces

agsh committed:
feat: brand new xml parsing style

agsh committed:
feat: backward and forward compatibility with the `any` elements in the `sequence` part 1

agsh committed:
chore: trying to figure it out the new workflow process #1

agsh committed:
Merge pull request #358 from agsh/v1-handle-xs-any

Add support for the xs:any tags for the different protocol extensions

agsh committed:
feat: media `setMetadataConfiguration` method

agsh committed:
feat: media `setAudioOutputConfiguration` method

agsh committed:
feat: media `setAudioDecoderConfiguration` method

agsh committed:
Merge pull request #361 from agsh/355-media1-methods

Media1 methods

agsh committed:
fix: media `getVideoSourceConfiguration` method fix

agsh committed:
Merge pull request #362 from agsh/355-media1-methods

355 media1 methods fix

agsh committed:
feat: new behavior for any tag in `linerase` function

agsh committed:
feat: fixes and tests for xs:any. Now working with this element works for any case

agsh committed:
Merge pull request #363 from agsh/v1-any-new-vision

feat: new behavior for any tag in `linerase` function

agsh committed:
fix: xs:any array-object behavior fix

agsh committed:
chore: update README

agsh committed:
chore: add `build` directory to test branch as a package

agsh committed:
fix: discovery new xml parser structure fix

agsh committed:
Merge pull request #368 from agsh/discovery-new-xml-parser-fix

fix: discovery new xml parser structure fix

agsh committed:
fix: discovery new xml parser structure fix build

agsh committed:
Merge branch 'master' into v1

# Conflicts:
#	.eslintrc.json
#	README.md
#	lib/device.js
#	package-lock.json
#	package.json
#	startServerMockup.js

agsh committed:
chore: update dependencies

agsh committed:
fix: eslint 9 linting

agsh committed:
chore: minimal node.js version now is 18.x

agsh committed:
fix: dns resolving for the tests

agsh committed:
Merge pull request #387 from agsh/v1-update-deps

chore: update dependencies

agsh committed:
feat: media2 getStreamUri method

agsh committed:
chore: run pipeline on push

agsh committed:
feat: media1 getStreamUri method with media2 call if supported (as in 0.x versions)

agsh committed:
feat: media1 tests

agsh committed:
feat: discovery tests

agsh committed:
feat: real discovery tests

agsh committed:
chore: remove build dir

agsh committed:
feat: discovery error tests

agsh committed:
Merge pull request #388 from agsh/v1-discovery-tests

feat: discovery error tests

agsh committed:
feat: getSnapshotUri

agsh committed:
Merge pull request #389 from agsh/v1-discovery-tests

feat: getSnapshotUri

agsh committed:
feat: setVideoEncoderConfiguration

agsh committed:
Merge pull request #390 from agsh/v1-discovery-tests

feat: setVideoEncoderConfiguration

agsh committed:
feat: setAudioEncoderConfiguration

agsh committed:
feat: setAudioSourceConfiguration

agsh committed:
feat: setMetadataConfiguration

agsh committed:
feat: setAudioOutputConfiguration

agsh committed:
feat: getVideoSourceConfigurationOptions

agsh committed:
Merge pull request #391 from agsh/v1-getVideoSourceConfigurationOptions

feat: getVideoSourceConfigurationOptions

agsh committed:
feat: getVideoSourceConfigurationOptions

agsh committed:
feat: multicast streaming

agsh committed:
feat: videosource modes get/set

agsh committed:
feat: OSD

agsh committed:
feat: OSD tests
chore: update HappyTimeSoft ONVIF Server

agsh committed:
chore: update ONVIF specs to 25.12

agsh committed:
feat: masks, optional profileToken

agsh committed:
feat: OSD, mask tests

agsh committed:
chore: upgrade HappyTimeSoft ONVIF server

agsh committed:
Merge pull request #392 from agsh/v1-lot-of-changes

V1 lot of changes

agsh committed:
feat: Mask tests

agsh committed:
Merge pull request #393 from agsh/v1-lot-of-changes

feat: Mask tests

agsh committed:
chore: update docs, media2 completed

agsh committed:
Merge pull request #394 from agsh/v1-lot-of-changes

chore: update docs, media2 completed

agsh committed:
fix: wrong naming for refs

agsh committed:
feat: blank for the events class,getEventProperties, getServiceCapabilities methods

agsh committed:
Merge pull request #395 from agsh/v1-events

V1 events

agsh committed:
fix: correct work with the same type names in different files

agsh committed:
Merge pull request #396 from agsh/v1-events

fix: correct work with the same type names in different files

agsh committed:
feat: events.pullMessages

agsh committed:
Merge pull request #397 from agsh/v1-events

feat: events.pullMessages

agsh committed:
feat: first implementation of the events

agsh committed:
Merge pull request #398 from agsh/v1-events

feat: first implementation of the events

agsh committed:
feat: full implementation of the events where we have one common subscription

agsh committed:
Merge pull request #400 from agsh/v1-events

feat: full implementation of the events where we have one common subscription

agsh committed:
feat: interfaces for pullMessagesResponse

agsh committed:
Merge pull request #401 from agsh/v1-events

feat: interfaces for pullMessagesResponse

agsh committed:
feat: structured Onvif-class events with typing

agsh committed:
feat: more event rules, event docs, trying to find out an error in the quick connection to the pull-point

agsh committed:
Merge pull request #402 from agsh/v1-events

V1 events

agsh committed:
feat: revision of media service, added getServiceCapabilities() method

agsh committed:
Merge pull request #403 from agsh/v1-media-finalize

feat: revision of media service, added getServiceCapabilities() method

agsh committed:
feat: startMulticastStreaming(), stopMulticastStreaming() methods

agsh committed:
Merge pull request #404 from agsh/v1-media-finalize

feat: startMulticastStreaming(), stopMulticastStreaming() methods

agsh committed:
feat: setSynchronizationPoint() method

agsh committed:
Merge pull request #405 from agsh/v1-media-finalize

feat: setSynchronizationPoint() method

agsh committed:
feat: getVideoSourceModes() method, videoSourceToken property

agsh committed:
Merge pull request #406 from agsh/v1-media-finalize

feat: getVideoSourceModes() method, videoSourceToken property

agsh committed:
chore: README update for the testing purposes

agsh committed:
Merge pull request #407 from agsh/v1-media-finalize-few-docs

chore: README update for the testing purposes

agsh committed:
chore: update README

agsh committed:
Merge pull request #408 from agsh/v1-media-finalize

V1 media finalize

agsh committed:
fix: getVideoSourceModes StringList interface

agsh committed:
Merge pull request #410 from agsh/v1-media-finalize

fix: getVideoSourceModes StringList interface

agsh committed:
feat: setVideoSourceMode() method

agsh committed:
feat: OSD methods

agsh committed:
Merge pull request #411 from agsh/v1-media-finalize

feat: setVideoSourceMode() method

agsh committed:
chore: working plan for PTZ

agsh committed:
Merge pull request #412 from agsh/v1-ptz

chore: working plan for PTZ

agsh committed:
feat: ptz.geoMove() method
feat: the body XML strings have been reworked for construction from the object.

agsh committed:
Merge pull request #413 from agsh/v1-ptz

feat: ptz.geoMove() method

agsh committed:
feat: ptz.getServiceCapabilities(), getNode() methods

agsh committed:
Merge pull request #414 from agsh/v1-ptz

feat: ptz.getServiceCapabilities(), getNode() methods

agsh committed:
feat: ptz.setConfiguration() method

agsh committed:
Merge pull request #415 from agsh/v1-ptz

feat: ptz.setConfiguration() method

agsh committed:
feat: ptz.sendAuxiliaryCommand() method

agsh committed:
Merge pull request #416 from agsh/v1-ptz

feat: ptz.sendAuxiliaryCommand() method

agsh committed:
feat: PTZ tour methods

agsh committed:
Merge pull request #418 from agsh/v1-ptz

feat: PTZ tour methods

agsh committed:
feat: all PTZ methods

agsh committed:
Merge pull request #419 from agsh/v1-ptz

feat: all PTZ methods

agsh committed:
feat: replay class

agsh committed:
Merge pull request #420 from agsh/v1-ptz

feat: replay class

agsh committed:
feat: Imaging class

agsh committed:
Merge pull request #421 from agsh/v1-imaging

feat: Imaging class

agsh committed:
feat: Recording class

agsh committed:
Merge pull request #422 from agsh/v1-recording

feat: Recording class

agsh committed:
feat: DoorControl class

agsh committed:
Merge pull request #423 from agsh/v1-doorcontrol

feat: DoorControl class

agsh committed:
chore: new interfaces build

agsh committed:
feat: Analytics module

agsh committed:
chore: README update

agsh committed:
Merge pull request #424 from agsh/v1-analytics

V1 analytics

agsh committed:
chore: service classes refactored

agsh committed:
Merge pull request #425 from agsh/v1-class-refactoring

chore: service classes refactored

RogerHardiman committed:
port splitargs from v0.x to v1 to fix parsing of Digest Authentication

RogerHardiman committed:
Fix bug in ContinuousPTZ

RogerHardiman committed:
Convert Example 3 (PTZ tests) to new library

RogerHardiman committed:
Merge pull request #427 from RogerHardiman/v1

port splitargs from v0.x to v1 to fix parsing of Digest Authentication

RogerHardiman committed:
Merge pull request #429 from RogerHardiman/ptz_fix

Fix bug in ContinuousPTZ

RogerHardiman committed:
Merge pull request #428 from RogerHardiman/convert_examples

Convert example 3 (PTZ) to use the new library

RogerHardiman committed:
Fix GetCapabilities

RogerHardiman committed:
Update Example 2 to use new library

RogerHardiman committed:
Merge pull request #430 from RogerHardiman/convert_examples

Fix GetCapabilities

RogerHardiman committed:
Merge pull request #431 from RogerHardiman/convert_examples

Update Example 2 to use new library

agsh committed:
Update README.md

agsh committed:
chore: splitArgs tests and fixes

agsh committed:
feat: DeviceIO class

agsh committed:
Merge pull request #433 from agsh/vi-deviceio

Vi deviceio

agsh committed:
feat: Display class

agsh committed:
Merge pull request #434 from agsh/v1-display

feat: Display class

agsh committed:
feat: Device class

agsh committed:
chore: remove old files

agsh committed:
Merge pull request #435 from agsh/v1-device-uncompleted

feat: Device class

agsh committed:
chore: fresh docs

agsh committed:
Merge pull request #436 from agsh/v1-docs-again

chore: fresh docs

agsh committed:
feat: ActionEngine class

agsh committed:
Merge pull request #437 from agsh/v1-actionengine

feat: ActionEngine class

agsh committed:
fix: events connection reset fix

agsh committed:
feat: event methods

agsh committed:
Merge pull request #438 from agsh/v1-events-fixes-and-upgrades

V1 events fixes and upgrades

agsh committed:
feat: events with notification producer

agsh committed:
feat: interfaces for the event filters

agsh committed:
feat: interfaces for the event filters

agsh committed:
feat: Subscription class

agsh committed:
fix: connection close for pulling

agsh committed:
feat: event error handling improvements and example

agsh committed:
chore: event error handling for pull-point sub tested on the real device

agsh committed:
feat: common event emitter for Onvif class

agsh committed:
fix: tests rewrote

agsh committed:
feat: real tests for global and filter events on the test server

agsh committed:
feat: WS-BaseNotification methods

agsh committed:
Merge pull request #439 from agsh/v1-event-filters

V1 event filters

agsh committed:
chore: ts 7 & lowered node.js version to 16

agsh committed:
feat: Thermal class

agsh committed:
feat: SHA-256 Digest auth

agsh committed:
chore: ts6

agsh committed:
Merge pull request #448 from agsh/v1-sha256

V1 sha256
